File: week13/11/3/P3.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

struct Compare {
    bool operator()(const pair<int, int>& a, const pair<int, int>& b) {
        // 첫 번째 기준: 첫 번째 요소 기준 오름차순
        if (a.first != b.first)
            return a.first > b.first;
        // 두 번째 기준: 두 번째 요소 기준 오름차순
        return a.second > b.second;
    }
};

class Solution {
public:
    vector<int> kWeakestRows(vector<vector<int>>& mat, int k) {
        vector<int> result;
        priority_queue<pair<int, int>, vector<pair<int, int>>, Compare> pq;
        for(int i = 0; i<mat.size(); i++){
            int count = 0;
            for (auto elem : mat[i]){
                if (elem == 1){
                    count++;
                }
                else{
                    break;
                }
            }
            pq.push(make_pair(count, i));
        }
        for(int j = 0; j < k;j++){
            result.push_back(pq.top().second);
            pq.pop();
        }
        return result;
    }
};

File: week13/11/3/P3.py
class Solution:
    def kWeakestRows(self, mat: List[List[int]], k: int) -> List[int]:
        result = []
        queue = []
        l = len(mat[0])
        for i in range(len(mat)):
            j = 0
            count = 0
            while j < l and mat[i][j] == 1:
                count += 1
                j += 1
            queue.append([count, i])
        queue.sort(key = lambda x : (x[0], x[1]))

        for rank in range(k):
            result.append(queue[rank][1])
        return result
